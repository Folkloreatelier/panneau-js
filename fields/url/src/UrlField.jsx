import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormGroup } from '@panneau/field';
import TextField from '@panneau/field-text';

import Url from './Url';

const propTypes = {
    name: PropTypes.string,
    label: PropTypes.string,
    value: PropTypes.string,
    schemes: PropTypes.arrayOf(PropTypes.string),
    onChange: PropTypes.func,
};

const defaultProps = {
    name: null,
    label: null,
    value: null,
    schemes: ['http://', 'https://', 'ftp://'],
    onChange: null,
};

class UrlField extends Component {
    constructor(props) {
        super(props);

        this.onChange = this.onChange.bind(this);

        this.url = new Url({
            schemes: props.schemes,
        });

        this.state = {
            scheme: this.url.getScheme(props.value),
        };
    }

    componentWillReceiveProps(nextProps) {
        const valueChanged = nextProps.value !== this.props.value;
        if (valueChanged) {
            const { value } = nextProps;
            this.setState({
                scheme: this.url.getScheme(value),
            });
        }
    }

    onChange(value) {
        const { scheme } = this.state;
        if (this.props.onChange) {
            this.props.onChange(this.url.withScheme(value, scheme));
        }
    }

    render() {
        const {
            label,
            name,
            value,
            ...other
        } = this.props;
        const { scheme } = this.state;

        return (
            <FormGroup
                className="form-group-url"
                name={name}
                label={label}
                {...other}
            >
                <TextField
                    inputOnly
                    prefix={scheme}
                    value={this.url.removeScheme(value)}
                    onChange={this.onChange}
                />
            </FormGroup>
        );
    }
}

UrlField.propTypes = propTypes;
UrlField.defaultProps = defaultProps;

export default UrlField;
