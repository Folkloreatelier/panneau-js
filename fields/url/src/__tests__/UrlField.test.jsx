/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import renderer from 'react-test-renderer';
// import { shallow } from 'enzyme';
// import sinon from 'sinon';
import UrlField from '../UrlField';

test('match snapshot', () => {
    const component = renderer.create((
        <UrlField
            label="Label"
            value="https://google.com"
        />
    ));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
