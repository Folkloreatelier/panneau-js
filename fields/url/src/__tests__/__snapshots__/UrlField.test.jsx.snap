// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div
  className="form-group form-group-url"
>
  <div
    className="form-group-label"
  >
    <label
      className="control-label"
      htmlFor={null}
    >
      Label
    </label>
  </div>
  <div
    className="form-group-inner"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      className="input-group"
    >
      <span
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
        >
          https://
        </span>
      </span>
      <input
        className="form-control field-text"
        disabled={false}
        id={null}
        name={null}
        onBlur={null}
        onChange={[Function]}
        onFocus={null}
        placeholder={null}
        type="text"
        value="google.com"
      />
    </div>
  </div>
</div>
`;
